data/transactions/logic/eval.go

```diff
@@ -2525,6 +2526,145 @@ func opEd25519verify(cx *EvalContext) {
+	pprev := prev - 1         // index of signature r
+	fourth := pprev - 1       // index of signature s
+	fifth := fourth - 1       // index of data
+
+	pkX := cx.stack[last].Bytes
+	pkY := cx.stack[prev].Bytes
+	sigR := cx.stack[pprev].Bytes
+	sigS := cx.stack[fourth].Bytes
+	msg := cx.stack[fifth].Bytes
+
+	var pubkey []byte
+	// if PK y component is set then interpret it as uncompressed PK
+	if len(pkY) != 0 {
+		x := new(big.Int).SetBytes(pkX)
+		y := new(big.Int).SetBytes(pkY)
+		pubkey = secp256k1.CompressPubkey(x, y)
+	} else {
+		// otherwise it is compressed PK
+		pubkey = pkX
+	}
+
+	signature := make([]byte, 0, len(sigR)+len(sigS))
+	signature = append(signature, sigR...)
+	signature = append(signature, sigS...)
+
+	var result bool
+
+	// secp256k1.VerifySignature easily panics on invalid input, catch it here and convert to an error
+	result = func() bool {
+		defer func() {
+			if x := recover(); x != nil {
+				cx.err = fmt.Errorf("ecdsa verify error: %v", x)
+				result = false
+			}
+		}()
+		return secp256k1.VerifySignature(pubkey, msg, signature)
+	}()
+
+	if result {
+		cx.stack[fifth].Uint = 1
+	} else {
+		cx.stack[fifth].Uint = 0
+	}
+	cx.stack[fifth].Bytes = nil
+	cx.stack = cx.stack[:fourth]
+}
+
+func opEcDsaPkDecompress(cx *EvalContext) {
+	ecdsaCurve := EcDsaCurve(cx.program[cx.pc+1])
+	fs, ok := ecDsaCurveSpecByField[ecdsaCurve]
+	if !ok || fs.version > cx.version {
+		cx.err = fmt.Errorf("invalid curve %d", ecdsaCurve)
+		return
+	}
+
+	if fs.field != Secp256k1 {
+		cx.err = fmt.Errorf("unsupported curve %d", fs.field)
+		return
+	}
+
+	last := len(cx.stack) - 1 // compressed PK
+
+	pubkey := cx.stack[last].Bytes
+	x, y := secp256k1.DecompressPubkey(pubkey)
+	if x == nil || y == nil {
+		cx.err = fmt.Errorf("invalid pubkey")
+		return
+	}
+
+	cx.stack[last].Uint = 0
+	cx.stack[last].Bytes = y.Bytes()
+
+	var sv stackValue
+	sv.Bytes = x.Bytes()
+	cx.stack = append(cx.stack, sv)
+}
+
+func opEcDsaPkRecover(cx *EvalContext) {
+	ecdsaCurve := EcDsaCurve(cx.program[cx.pc+1])
+	fs, ok := ecDsaCurveSpecByField[ecdsaCurve]
+	if !ok || fs.version > cx.version {
+		cx.err = fmt.Errorf("invalid curve %d", ecdsaCurve)
+		return
+	}
+
+	if fs.field != Secp256k1 {
+		cx.err = fmt.Errorf("unsupported curve %d", fs.field)
+		return
+	}
+
+	last := len(cx.stack) - 1 // index of signature r
+	prev := last - 1          // index of signature s
+	pprev := prev - 1         // recover id
+	fourth := pprev - 1       // index of data
+
+	sigR := cx.stack[last].Bytes
+	sigS := cx.stack[prev].Bytes
+	recid := cx.stack[pprev].Uint
+	msg := cx.stack[fourth].Bytes
+
+	if recid > 3 {
+		cx.err = fmt.Errorf("invalid recovery id: %d", recid)
+		return
+	}
+
+	signature := make([]byte, 0, len(sigR)+len(sigS)+1)
+	signature = append(signature, sigR...)
+	signature = append(signature, sigS...)
+	signature = append(signature, uint8(recid))
+
+	pk, err := secp256k1.RecoverPubkey(msg, signature)
```

Comment:
The returned `pk` is an [uncompressed public key](https://github.com/ethereum/go-ethereum/blob/d8ff53dfb8a516f47db37dbc7fd7ad18a1e8a125/crypto/secp256k1/ext.h#L34), and the description for this op says it returns a compressed public key. So you could do `x, y := secp256k1.S256().Unmarshal(pk)` to extract the x and y components to return those.

Returning an uncompressed key is useful for this function because Ethereum addresses are derived from the keccak hash of `x` concat `y`.

Reply:
returning x, y components now